components/engine_traits/src/tablet.rs

```diff
@@ -0,0 +1,398 @@
+    /// likely the region Id, the suffix could be the current raft log
+    /// index. The reason to have suffix is that we can keep more than one
+    /// tablet for a region.
+    fn open_tablet(&self, _id: u64, _suffix: Option<u64>, _path: &Path) -> Result<EK> {
+        Ok(self.tablet.clone())
+    }
+
+    /// Destroy the tablet and its data
+    fn destroy_tablet(&self, _id: u64, _suffix: Option<u64>, _path: &Path) -> Result<()> {
+        Ok(())
+    }
+
+    /// Check if the tablet with specified path exists
+    fn exists(&self, _path: &Path) -> bool {
+        true
+    }
+}
+
+/// A global registry for all tablets.
+struct TabletRegistryInner<EK> {
+    // region_id, suffix -> tablet
+    tablets: Mutex<HashMap<u64, CachedTablet<EK>>>,
+    tombstone: Mutex<Vec<(u64, u64)>>,
+    factory: Box<dyn TabletFactory<EK>>,
+    root: PathBuf,
+}
+
+pub struct TabletRegistry<EK> {
+    // One may consider to add cache to speed up access. But it also makes it more
+    // difficult to gc stale cache.
+    tablets: Arc<TabletRegistryInner<EK>>,
+}
+
+impl<EK> Clone for TabletRegistry<EK> {
+    fn clone(&self) -> Self {
+        Self {
+            tablets: self.tablets.clone(),
+        }
+    }
+}
+
+unsafe impl<EK: Send> Send for TabletRegistry<EK> {}
+unsafe impl<EK: Send> Sync for TabletRegistry<EK> {}
+
+impl<EK> TabletRegistry<EK> {
+    pub fn new(factory: Box<dyn TabletFactory<EK>>, path: impl Into<PathBuf>) -> Result<Self> {
+        let root = path.into();
+        std::fs::create_dir_all(&root)?;
+        Ok(TabletRegistry {
+            tablets: Arc::new(TabletRegistryInner {
+                tablets: Mutex::new(HashMap::default()),
+                factory,
+                root,
+                tombstone: Mutex::default(),
+            }),
+        })
+    }
+
+    pub fn tablet_name(&self, prefix: &str, id: u64, suffix: u64) -> String {
+        format!("{}{}_{}", prefix, id, suffix)
+    }
+
+    pub fn tablet_root(&self) -> &Path {
+        &self.tablets.root
+    }
+
+    pub fn tablet_path(&self, id: u64, suffix: u64) -> PathBuf {
+        let name = self.tablet_name("", id, suffix);
+        self.tablets.root.join(name)
+    }
+
+    /// Gets a tablet.
+    pub fn get(&self, id: u64) -> Option<CachedTablet<EK>>
+    where
+        EK: Clone,
+    {
+        let tablets = self.tablets.tablets.lock().unwrap();
+        tablets.get(&id).cloned()
+    }
+
+    /// Gets a tablet, create a default one if it doesn't exist.
+    pub fn get_or_default(&self, id: u64) -> CachedTablet<EK>
+    where
+        EK: Clone,
+    {
+        let mut tablets = self.tablets.tablets.lock().unwrap();
+        tablets
+            .entry(id)
+            .or_insert_with(|| CachedTablet::new(None))
+            .clone()
+    }
+
+    pub fn tablet_factory(&self) -> &dyn TabletFactory<EK> {
+        self.tablets.factory.as_ref()
+    }
+
+    pub fn remove(&self, id: u64) {
```

Comment:
Why not handle the whole deletion at this level? (instead of calling `remove && destroy_tablet`). In fact, exposing the factory in multi-rocks code looks bug-prone to me.

Reply:
Because tablet can't be destroyed immediately. It should wait till no one is referencing the tablet. In complex case like split/merge, it should also wait till new tablet is persistent.